src/Avalonia.Base/StyledElement.cs

```diff
@@ -886,70 +895,81 @@ private void ClearLogicalParent(IList children)
             }
         }
 
-        private void DetachStyles(IReadOnlyList<StyleBase>? styles = null)
+        private void DetachStyles(IReadOnlyList<Style> styles)
         {
-            var valueStore = GetValueStore();
+            var values = GetValueStore();
+            values.BeginStyling();
+            try { values.RemoveFrames(styles); }
+            finally { values.EndStyling(); }
 
-            valueStore.BeginStyling();
-
-            for (var i = valueStore.Frames.Count - 1; i >= 0; --i)
+            if (_logicalChildren is not null)
             {
-                if (valueStore.Frames[i] is StyleInstance si &&
-                    (styles is null || styles.Contains(si.Source)))
+                var childCount = _logicalChildren.Count;
+
+                for (var i = 0; i < childCount; ++i)
                 {
-                    valueStore.RemoveFrame(si);
+                    (_logicalChildren[i] as StyledElement)?.DetachStyles(styles);
                 }
             }
-
-            valueStore.EndStyling();
-            _styled = false;
         }
 
-        private void InvalidateStylesOnThisAndDescendents()
+        private void NotifyResourcesChanged(
+            ResourcesChangedEventArgs? e = null,
+            bool propagate = true)
         {
-            InvalidateStyles();
-
-            if (_logicalChildren is object)
+            if (ResourcesChanged is object)
             {
-                var childCount = _logicalChildren.Count;
+                e ??= ResourcesChangedEventArgs.Empty;
+                ResourcesChanged(this, e);
+            }
 
-                for (var i = 0; i < childCount; ++i)
-                {
-                    (_logicalChildren[i] as StyledElement)?.InvalidateStylesOnThisAndDescendents();
-                }
+            if (propagate)
+            {
+                e ??= ResourcesChangedEventArgs.Empty;
+                NotifyChildResourcesChanged(e);
             }
         }
 
-        private void DetachStylesFromThisAndDescendents(IReadOnlyList<StyleBase> styles)
+        private static IReadOnlyList<Style>? FlattenStyles(IReadOnlyList<IStyle> styles)
```

Comment:
I am pretty sure we can return a pooled list from here and dispose after "RemoveFrames" was called.